README.md
# Hydra adaptation
Very fast network logon cracker

debian/changelog
hydra (8.1-1~kali1) kali; urgency=medium

  * Import new version from debian

 -- Sophie Brun <sophie@freexian.com>  Tue, 10 Feb 2015 15:57:34 +0100

hydra (8.1-1) unstable; urgency=medium

  * New upstream release.
  * Bump Standards-Version to 3.9.6 (no changes).
  * Refresh 02_using_supported_protocols.diff and
    08_fix_spelling_mistakes.diff patches.
  * Remove d/source/options, now xz format is default.
  * Update 08_fix_spelling_mistakes.diff patch.
  * d/copyright: Add missing License values.

 -- Julián Moreno Patiño <julian@debian.org>  Wed, 10 Dec 2014 07:33:34 -0500

hydra (8.0-1kali1) kali; urgency=low

  * Upstream import

 -- Mati Aharoni <muts@kali.org>  Fri, 23 May 2014 04:52:27 -0400

hydra (8.0-1) unstable; urgency=medium

  * New upstream release. (LP: #1275463)
  * debian/control
    + Bump Standards-Version to 3.9.5 (no changes).
    + Add dh-exec in B-D
    + Add wget | curl to Recommends field
  * debian/patches
    + Refresh all patches
    + Add 09_use_correct_install_dir_dlp4hydra.diff
      + Change upstream path for one suitable for GNU Debian
  * debian/copyright
    + Extend copyright holders years
    + Add dpl4hydra.sh copyright
  * Remove debian/source/options
    + Use xz compression.
  * debian/rules
    + Remove override_dh_builddeb
      + Now dpkg-deb uses xz compression by default
  * debian/hydra.manpages
    + Add dpl4hydra.1 file
    + Add hydra-wizard.1 file
  * Add hydra.install file
    + Include two new scripts
      + dpl4hydra: Generate a default password list
      + hydra-wizard: A wizard script to use Hydra

 -- Daniel Echeverry <epsilon77@gmail.com>  Fri, 30 May 2014 22:05:44 -0500

hydra (7.6-0kali1) kali; urgency=low

  * Upstream import

 -- Mati Aharoni <muts@kali.org>  Sat, 28 Dec 2013 10:34:01 -0500

hydra (7.5-2) unstable; urgency=low

  * debian/control
    + Remove libncp-dev in B-D. (Closes: #728864)
  * debian/patches
    + Refresh 02_using_supported_protocols.diff patch.
      + Remove the support of ncp protocol.

 -- Daniel Echeverry <epsilon77@gmail.com>  Fri, 15 Nov 2013 22:02:21 -0500

hydra (7.5-1) unstable; urgency=low

  * New upstream release
  * debian/copyright
    + Move the license from GPL-3.0+ to AGPL-3.0 with OpenSSL exception
    + Extend copyright holders years
  * debian/patches
    + Refreshed patch 02_using_supported_protocols.diff
    + Refreshed patch 05_add_option_to_cflag.diff
    + Refreshed patch 07_remove_troubled_files.diff
    + Refreshed patch 08_fix_spelling_mistakes.diff

 -- Daniel Echeverry <epsilon77@gmail.com>  Sat, 31 Aug 2013 22:28:57 -0500

hydra (7.4.2-2) unstable; urgency=low

  * Use ${binary:Version} in hydra-gtk package
    instead bump manually the version.
  * Disable libafpclient-dev support due to #702889
    + Update 02_using_supported_protocols.diff patch to remove
      the support of afp protocol.
  * Use canonical URIs for Vcs-* fields.
  * Add Keywords entry to d/hydra-gtk.desktop.

 -- Julián Moreno Patiño <julian@debian.org>  Mon, 15 Jul 2013 15:44:45 -0500

hydra (7.4.2-1) unstable; urgency=low

  * New upstream release.

 -- Daniel Echeverry <epsilon77@gmail.com>  Sat, 12 Jan 2013 09:23:51 -0500

hydra (7.4-1) unstable; urgency=low

  * New upstream release.
  * d/watch: Update regular expression.
  * d/control:
    + Remove DM-Upload-Allowed.
    + Update my email.
    + Add libncurses-dev to B-D.
    + Improve description. Thanks to Justin B Rye. (Closes: #680540)
    + Bump Standards-Version to 3.9.4 (no changes).
  * d/copyright: Update my email.
  * d/patches:
    + Refresh 02_using_supported_protocols.diff patch offset.
    + Update 04_fix_hyphen-used-as-minus-sign.diff.
    + Refresh 05_add_option_to_cflag.diff patch offset.
    + Add 08_fix_spelling_mistakes.diff patch to fix spelling
      mistakes.

 -- Julián Moreno Patiño <julian@debian.org>  Sun, 23 Dec 2012 14:22:46 -0500

hydra (7.3-1) unstable; urgency=low

  * New upstream release.
  * Remove 04_fix_spelling_error.diff patch
    + Merge with upstream.
  * Remove some lines in 05_add_option_to_cflag.diff patch.
    + Merge with upstream.
    + Improve DEP3 Description.
  * Refresh 07_remove_troubled_files.diff patch offset.
  * Add 04_fix_hyphen-used-as-minus-sign.diff patch to
    fix hyphen-used-as-minus-sign lintian warning in the
    manpage.

 -- Julián Moreno Patiño <darkjunix@gmail.com>  Wed, 23 May 2012 21:25:24 -0500

hydra (7.2-2) unstable; urgency=low

  * Update 05_add_option_to_cflag.diff patch to use properly
    Hardening flags. (Closes: #665281)
    + Thanks to Simon Ruderich for the patch.
  * Bump debhelper to 9 in B-D.
    + Remove debian/source/lintian-overrides to do lintian happy
      about package-needs-versioned-debhelper-build-depends 9.
  * Bump Standards-Version to 3.9.3.
    + Update to DEP5 copyright format 1.0.

 -- Daniel Echeverry <epsilon77@gmail.com>  Sat, 24 Mar 2012 23:51:36 -0500

hydra (7.2-1) unstable; urgency=low

  * New upstream release.
  * Refresh 05_add_option_to_cflag.diff patch offset.
  * Extend copyright holders years.
    + Add missing copyright.

  [Julián Moreno Patiño]
  * Add 07_remove_troubled_files.diff patch to avoid
    a FTBFS in a second row.
  * Bump debhelper to 8.1.3. (compat level 9).
  * Update 02_using_supported_protocols.diff patch to add
    support of AFP protocol to Linux and KfreeBSD arches.
    + Add libafpclient-dev[!hurd-any] to B-D.

 -- Daniel Echeverry <epsilon77@gmail.com>  Thu, 16 Feb 2012 23:32:00 -0500

hydra (7.1-2) unstable; urgency=low

  * Switch compat level 8 to 9 to use hardening flags.
  * Update 05_add_option_to_cflag.diff patch
    to use properly Hardening flags.
  * Add 06_show_xhydra_build_output.diff patch
    to show verbose build mode in xhydra.
  * Remove unnecessary newline in debian/rules.
  * Use firebird-dev instead firebird2.5-dev in B-D. (Closes: #653377)
  * Add lintian override to fix
    package-needs-versioned-debhelper-build-depends
    warning for experimental debhelper version.


hydra-cvs.c
#include "hydra-mod.h"

extern int hydra_data_ready_timed(int socket, long sec, long usec);

extern char *HYDRA_EXIT;
char *buf;

int start_cvs(int s, char *ip, int port, unsigned char options, char *miscptr, FILE * fp) {
  char *empty = "";
  char *login, *pass, buffer[1024], pass2[513];
  int i;
  char *directory = miscptr;

/* evil cvs encryption sheme... 
        0 111           P 125           p  58
! 120   1  52   A  57   Q  55   a 121   q 113
"  53   2  75   B  83   R  54   b 117   r  32
        3 119   C  43   S  66   c 104   s  90
        4  49   D  46   T 124   d 101   t  44
% 109   5  34   E 102   U 126   e 100   u  98
&  72   6  82   F  40   V  59   f  69   v  60
' 108   7  81   G  89   W  47   g  73   w  51
(  70   8  95   H  38   X  92   h  99   x  33
)  64   9  65   I 103   Y  71   i  63   y  97
*  76   : 112   J  45   Z 115   j  94   z  62
+  67   ;  86   K  50           k  93
, 116   < 118   L  42           l  39
-  74   = 110   M 123           m  37
.  68   > 122   N  91           n  61
/  87   ? 105   O  35   _  56   o  48
*/

  char key[] = { 0, 120, 53, 0, 0, 109, 72, 108, 70, 64, 76, 67, 116, 74, 68, 87,
    111, 52, 75, 119, 49, 34, 82, 81, 95, 65, 112, 86, 118, 110, 122, 105,
    0, 57, 83, 43, 46, 102, 40, 89, 38, 103, 45, 50, 42, 123, 91, 35,
    125, 55, 54, 66, 124, 126, 59, 47, 92, 71, 115, 0, 0, 0, 0, 56,
    0, 121, 117, 104, 101, 100, 69, 73, 99, 63, 94, 93, 39, 37, 61, 48,
    58, 113, 32, 90, 44, 98, 60, 51, 33, 97, 62
  };

  if (strlen(login = hydra_get_next_login()) == 0)
    login = empty;
  if (strlen(pass = hydra_get_next_password()) == 0)
    pass = empty;

  memset(pass2, 0, sizeof(pass2));
  strncpy(pass2, pass, 512);

  for (i = 0; i < strlen(pass); i++) {
    pass2[i] = key[pass2[i] - 0x20];
  }

  snprintf(buffer, sizeof(buffer), "BEGIN VERIFICATION REQUEST\n%s\n%s\nA%s\nEND VERIFICATION REQUEST\n", directory, login, pass2);

  i = 57 + strlen(directory) + strlen(login) + strlen(pass2);

  if (hydra_send(s, buffer, i - 1, 0) < 0) {
    return 1;
  }

  if (hydra_data_ready_timed(s, 5, 0) > 0) {
    buf = hydra_receive_line(s);
    if (strstr(buf, "I LOVE YOU\n")) {
      hydra_report_found_host(port, ip, "cvs", fp);
      hydra_completed_pair_found();
      free(buf);
      if (memcmp(hydra_get_next_pair(), &HYDRA_EXIT, sizeof(HYDRA_EXIT)) == 0) {
        return 3;
      }
    } else if (strstr(buf, "no such user") || strstr(buf, "E PAM start error: Critical error - immediate abort\n")) {
      if (verbose) {
        hydra_report(stderr, _("[VERBOSE] User %s does not exist\n"), login);
      }
      hydra_completed_pair_skip();
      free(buf);
      if (memcmp(hydra_get_next_pair(), &HYDRA_EXIT, sizeof(HYDRA_EXIT)) == 0) {
        return 3;
      }
    }
    /* "I HATE YOU\n" case */
    free(buf);
    return 3;
  }

  return 3;
}

void service_cvs(char *ip, int sp, unsigned char options, char *miscptr, FILE * fp, int port) {
  int run = 1, next_run = 1, sock = -1;
  int myport = PORT_CVS, mysslport = PORT_CVS_SSL;

  hydra_register_socket(sp);

  if ((miscptr == NULL) || (strlen(miscptr) == 0)) {
    miscptr = "/root";
  }

  while (1) {
    if (memcmp(hydra_get_next_pair(), &HYDRA_EXIT, sizeof(HYDRA_EXIT)) == 0)
      return;

    switch (run) {
    case 1:                    /* connect and service init function */
      if (sock >= 0)
        sock = hydra_disconnect(sock);

//      usleep(300000);
      if ((options & OPTION_SSL) == 0) {
        if (port != 0)
          myport = port;
        sock = hydra_connect_tcp(ip, myport);
        port = myport;
      } else {
        if (port != 0)
          mysslport = port;
        sock = hydra_connect_ssl(ip, mysslport);
        port = mysslport;
      }

      if (sock < 0) {
        hydra_report(stderr, _("[ERROR] Child with pid %d terminating, can not connect\n"), (int) getpid());
        hydra_child_exit(1);
      }
      next_run = start_cvs(sock, ip, port, options, miscptr, fp);
      break;
    case 3:                    /* clean exit */
      if (sock >= 0)
        sock = hydra_disconnect(sock);
      hydra_child_exit(2);
      return;
    default:
      hydra_report(stderr, _("[ERROR] Caught unknown return code, exiting!\n"));
      hydra_child_exit(2);
    }
    run = next_run;
  }
}

int service_cvs_init(char *ip, int sp, unsigned char options, char *miscptr, FILE * fp, int port) {
  // called before the childrens are forked off, so this is the function
  // which should be filled if initial connections and service setup has to be
  // performed once only.
  //
  // fill if needed.
  // 
  // return codes:
  //   0 all OK
  //   -1  error, hydra will exit, so print a good error message here

  return 0;
}
